File: 2026-03/2026-03-08/hslValidator.js
function isValidHSL(hsl) {
  const matchResult = /^hsl\( *(?<hue>\d{1,3}) *, *(?<saturation>\d{1,3})% *, *(?<lightness>\d{1,3})% *\) *;?$/.exec(hsl)?.groups;
  
  if (matchResult 
    && 0 <= matchResult.hue && matchResult.hue <= 360
    && 0 <= matchResult.saturation && matchResult.saturation <= 100
    && 0 <= matchResult.lightness && matchResult.lightness <= 100
  )
    return true;
  else
    return false;
}

// --- TEST SUITE ---

const testsText = String.raw`
1. isValidHSL("hsl(240, 50%, 50%)") should return true.
2. isValidHSL("hsl( 200 , 50% , 75% )") should return true.
3. isValidHSL("hsl(99,60%,80%);") should return true.
4. isValidHSL("hsl(0, 0%, 0%) ;") should return true.
5. isValidHSL("hsl(  10  ,  20%   ,  30%   )    ;") should return true.
6. isValidHSL("hsl(361, 50%, 80%)") should return false.
7. isValidHSL("hsl(300, 101%, 70%)") should return false.
8. isValidHSL("hsl(200, 55%, 75)") should return false.
9. isValidHSL("hsl (80, 20%, 10%)") should return false.
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);

    const comparison = Array.isArray(testOutput)
      ? arraysEqual(functionCallOutput, testOutput)
      : functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      ++failCount;
    }
  })

  console.log("----------------------------",
    failCount
    ? `\n⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "\n🎉SUCCESS: All tests PASSED."
  );
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (let i = 0; i < a.length; ++i) {
    if (Array.isArray(a[i])) {
      if (Array.isArray(b[i])) {
        if (!arraysEqual(a[i], b[i]))
          return false;
      }
      else 
        return false;
    }

    else if (a[i] !== b[i])
      return false;
  }

  return true;
}

runTests(testData);

File: 03 March/03-08/hsl_validator.py
def is_valid_hsl(hsl):
    match_result = re.match(r"^hsl\( *(?P<hue>\d{1,3}) *, *(?P<saturation>\d{1,3})% *, *(?P<lightness>\d{1,3})% *\) *;?$", hsl)

    if not match_result:
        return False

    groups_dict = match_result.groupdict()
    hue = int(groups_dict['hue'])
    saturation = int(groups_dict['saturation'])
    lightness = int(groups_dict['lightness'])
    
    if 0 <= hue <= 360 and 0 <= saturation <= 100 and 0 <= lightness <= 100:
        return True
    else:
        return False
        

# --- TEST SUITE ---

tests_text = r'''
1. is_valid_hsl("hsl(240, 50%, 50%)") should return True.
2. is_valid_hsl("hsl( 200 , 50% , 75% )") should return True.
3. is_valid_hsl("hsl(99,60%,80%);") should return True.
4. is_valid_hsl("hsl(0, 0%, 0%) ;") should return True.
5. is_valid_hsl("hsl(  10  ,  20%   ,  30%   )    ;") should return True.
6. is_valid_hsl("hsl(361, 50%, 80%)") should return False.
7. is_valid_hsl("hsl(300, 101%, 70%)") should return False.
8. is_valid_hsl("hsl(200, 55%, 75)") should return False.
9. is_valid_hsl("hsl (80, 20%, 10%)") should return False.
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

    print("----------------------------",
        f"\n⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "\n🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)